core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -1109,21 +1084,38 @@ private boolean shouldPerformTransactionOperation(StoreStatus storeStatus, Predi
                               });
 
                         AtomicLong writeCount = new AtomicLong();
-                        entryWriteFlowable = entryWriteFlowable.doOnNext(me -> writeCount.incrementAndGet());
 
                         Flowable<NonBlockingStore.SegmentedPublisher<MarshallableEntry<K, V>>> writeFlowable;
 
                         if (segmented) {
+                           // Note the writeCount includes entries that aren't written due to being shared
+                           // at this point
+                           entryWriteFlowable = entryWriteFlowable.doOnNext(obj -> writeCount.incrementAndGet());
                            writeFlowable = entryWriteFlowable
                                  .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
                                  .map(SegmentPublisherWrapper::wrap);
+                           if (shared) {
+                              // The writeCount will be decremented for each grouping of values ignored
+                              writeFlowable = filterSharedSegments(writeFlowable, writeCount);
+                           }
                         } else {
+                           if (shared) {
+                              entryWriteFlowable = entryWriteFlowable.filter(me ->
+                                    distributionManager.getCacheTopology().getDistribution(me.getKey()).isPrimary());
+                           }
+                           entryWriteFlowable = entryWriteFlowable.doOnNext(obj -> writeCount.incrementAndGet());
                            writeFlowable = Flowable.just(SingleSegmentPublisher.singleSegment(entryWriteFlowable));
                         }
 
                         if (trace) {
-                           removeFlowable = removeFlowable.doOnSubscribe(sub ->
+                           writeFlowable = writeFlowable.doOnSubscribe(sub ->
                               log.tracef("Store %s has subscribed to write batch", storeStatus.store));
+                           writeFlowable = writeFlowable.map(sp -> {
+                              int segment = sp.getSegment();
+                              return SingleSegmentPublisher.singleSegment(segment, Flowable.fromPublisher(sp)
+                                    .doOnNext(me -> log.tracef("Emitting entry %s for write to segment %s",
+                                          me, segment)));
+                           });
                         }
 
                         return Single.fromCompletionStage(flowableHandler.handleFlowables(storeStatus.store(),
```

Comment:
It's not from this PR, but can we break this into a `CompletionStage<Void>` var that is passed to `fromCompletionStage`. It is very difficult to read atm.

Reply:
Okay.